src/typescript-worker/typescript.ts

```diff
@@ -145,19 +169,73 @@ export class TypeScriptBuilder {
   }
 }
 
+interface VersionedFile {
+  version: number;
+  content: string;
+}
+
 class WorkerLanguageServiceHost implements ts.LanguageServiceHost {
   readonly compilerOptions: ts.CompilerOptions;
   readonly packageRoot: string;
-  readonly files: Map<string, string>;
+  readonly files: Map<string, VersionedFile>;
 
   constructor(
-    files: Map<string, string>,
     packageRoot: string,
-    compilerOptions: ts.CompilerOptions
+    compilerOptions: ts.CompilerOptions,
+    files?: Map<string, VersionedFile>
   ) {
     this.packageRoot = packageRoot;
     this.compilerOptions = compilerOptions;
-    this.files = files;
+    this.files = files || new Map();
+  }
+
+  /*
+   *  When a new new "process" command is received, we iterate through all of the files,
+   *  and update files accordingly depending on if they have new content or not.
+   *
+   *  With how the TS API works, we can use simple versioning to tell the
+   *  Language service that a file has been updated
+   *
+   *  If the file submitted is a new file, we add it to our collection
+   */
+  updateFileContentIfNeeded(fileName: string, content: string) {
+    const file = this.files.get(fileName);
+    if (file && this._fileContentHasUpdated(file, content)) {
+      file.content = content;
+      file.version += 1;
+    } else {
+      this.addFile(fileName, {content, version: 0});
+    }
+  }
+
+  /**
+   * Sync up the freshly acquired project files.
+   * In the syncing process files yet to be added are added, and versioned.
+   * Files that existed already but are modified are updated, and their version number
+   * gets bumped fo that the languageservice knows to update these files.
+   * */
+  sync(files: Map<string, string>) {
+    files.forEach((file, fileName) =>
+      this.updateFileContentIfNeeded(fileName, file)
+    );
```

Comment:
What happens if a file is deleted? Maybe we should iterate over the current `files`, and remove the ones that don't exist anymore?

I wonder if that will work properly in this edge case?:

1. compile with foo at version 0
2. delete foo and compile without it
3. re-create foo with different content, but it gets version 0 again because its new

Reply:
Oh good point! I'll need to cover the deletion case. 

.. What comes to the deletion and re-addition, I'll need to check out if the LS is capable of handling it out of the box, or if we need to add some flags etc.